src/core/sock/sockinfo_tcp.cpp

```diff
@@ -4556,6 +4556,42 @@ int sockinfo_tcp::tcp_setsockopt(int __level, int __optname, __const void *__opt
             ret = -1;
             errno = EINVAL;
             break;
+        case SO_XLIO_ISOLATE:
+            pass_to_os_cond = false;
+            if (__optlen == sizeof(int)) {
+                if (!m_p_connected_dst_entry && m_rx_ring_map.empty()) {
+                    int tempval = *reinterpret_cast<const int *>(__optval);
+                    bool ring_isolated =
+                        m_ring_alloc_log_rx.get_ring_alloc_logic() == RING_LOGIC_ISOLATE &&
+                        m_ring_alloc_log_tx.get_ring_alloc_logic() == RING_LOGIC_ISOLATE;
+
+                    if (tempval == SO_XLIO_ISOLATE_DEFAULT && !ring_isolated) {
+                        /* Do nothing. */
+                        break;
+                    }
+                    if (tempval == SO_XLIO_ISOLATE_SAFE) {
+                        if (safe_mce_sys().tcp_ctl_thread ==
+                                option_tcp_ctl_thread::CTL_THREAD_DELEGATE_TCP_TIMERS &&
+                            !ring_isolated) {
+                            /* XXX TODO Is it correct way? */
```

Comment:
What are the requirements for correct or incorrect?
How can we validate/test for "correctness"?

Reply:
There are 2 requirements here:
1. We need to replace one lock object with another. From C++ perspective, is it optimal way?
2. We must (a) serialize lock replacement with the previous lock usage and (b) guarantee that the previous lock is not locked.

2(a) and 2(b) are guaranteed if user calls setsockopt() from the native thread (taking into account that this is branch for "TCP delegated timers").

Reply:
It seems that this operation is inherently unsafe.
The locking and unlocking is a two-step operation:

1. unique_ptr dereference
2. locking the dereferenced lock pointer

The problematic scenario may happen when:

- Two threads may concurrently de-reference the pointer, and one may obtain the lock and replace the pointer.
- After the first thread releases/replaces the lock, the second thread may attempt to use a dangling lock pointer.

The **solution** might require a lock-replacing lock. To make the lock swapping a thread-safe operation, we might need the `mulilock::lock` operation to obtain the lock-replacing lock as well. The performance impact might be significant.